### PR TITLE
Remove show-build-info command and generate build-info as a side-effect of 'build'

### PR DESCRIPTION
**Pull request by fendor**
This is the first step for implementing Approach 3 of #7489.
Mainly extracted from #7478 

In particular, add a flag `--enable-build-info` to configure of `Setup.hs` and generate build-info.json right next to `setup-config`.

Readable by tooling.

* [x] Cabal Tests using `build-type: Simple`
* [x] Cabal Tests using `build-type: Custom`
* [x] Optional: Add `cabal configure --enable-dump-buildinfo` (e.g. enable/disable the generation)
* [x] Add `buildInfoJson` to `plan.json` output
* [x] Guard against older Cabal versions

### CODE REVIEW

File: cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs
Comment (fendor) [inline cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs:1092]: This is intentionally terrible so someone can tell me how to improve it and where it should be located

File: Cabal/src/Distribution/Simple/BuildPaths.hs
Comment (fendor) [inline Cabal/src/Distribution/Simple/BuildPaths.hs:72]: Possible inconsistency, file name is `build-info` but other flags are `buildinfo`. Any preferences?

Comment (Mikolaj) [reply Cabal/src/Distribution/Simple/BuildPaths.hs:72]: As far as I'm concerned buildinfo is a word.

File: Cabal/src/Distribution/Simple/Setup.hs
Comment (hasufell) [inline Cabal/src/Distribution/Simple/Setup.hs:567]: Why conditionally? Is there a reason not to write it always?

Comment (fendor) [reply Cabal/src/Distribution/Simple/Setup.hs:567]: I did not want to change the default behaviour, but it would be indeed better if it was always generated. Subject to bikeshed.

Comment (jneira) [reply Cabal/src/Distribution/Simple/Setup.hs:567]: performance is not a problem I suppose

Comment (DanielG) [reply Cabal/src/Distribution/Simple/Setup.hs:567]: I woudn't be so sure that performance couldn't be a problem, best to err on the side of caution when people with pitchforks are sure to follow if `build` speed regresses ;)

If it turns out to be unproblematic it's an easy default to change: just ignore the flag going forward, but the other way around isn't so easy since tooling would break if we introduce a flag.

Lastly my understanding is that hie-bios currently always does at least one (implicit) build call anyway though the repl command so we're just maintaining the status quo if it has to do one with a special flag to get the buildinfo.

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (DanielG) [inline cabal-install/src/Distribution/Client/ProjectPlanOutput.hs:154]: IIRC we decided (off-list) this should be nullable to represent build-type:Custom with lib:Cabal being too old.

**Review by DanielG — COMMENTED** [review]
This is looking pretty good. I'm super excited that we figured out how to roll this into build rather than introducing a new command.

Apart from nits and bikeshedding the only real points that need addressing are the non-nullability of the compiler `"path"`, the nullability of `"build-info"` in plan.json or an alternate mechanism for downstreams to detect if the "lib:Cabal too old" case is happening and as always moar docs ;)

File: Cabal/src/Distribution/Simple/BuildTarget.hs
Comment (DanielG) [inline Cabal/src/Distribution/Simple/BuildTarget.hs:67]: [[GH doesn't let me review the commit message so this is here instead]]

Seems this commit doesn't do much of anything anymore the commit message is out of sync here too.

File: Cabal/Cabal.cabal
Comment (DanielG) [inline Cabal/Cabal.cabal:258]: [[GH doesn't let me review the commit message so this is here instead]]

> introduces the buildinfo-components-only flag

I don't see this being introduced here, seems like the commit message has fallen out of sync with the diff?

File: Cabal/src/Distribution/Simple/Utils/Json.hs
Comment (DanielG) [inline Cabal/src/Distribution/Simple/Utils/Json.hs]: Would have been nice to do the refactor and move of this module in separate commits to more easily see what changed.

**Review by DanielG — COMMENTED** [review]
Just another per-commit review pass. Something seems off about the commit messages not being in sync with the diffs.

File: Cabal/src/Distribution/Simple/BuildTarget.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/BuildTarget.hs:67]: Remnants of rebasing, removing changes, adding changes, etc... in the hopes to have some kind of commit history. However, it seems currently better to drop and squash commits that don't do anything anymore.

File: Cabal/src/Distribution/Simple/Utils/Json.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/Utils/Json.hs]: The json changes have been extracted to https://github.com/haskell/cabal/pull/7477, this PR will be rebased on top of it once it has been merged.

File: Cabal/src/Distribution/Simple/Setup.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/Setup.hs:567]: Indeed better, will update

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:5]: no flag `--dump-buildinfo` exists yet, I would leave that to a follow-up PR, unless someone objects

File: Cabal/src/Distribution/Simple/Setup.hs
Comment (jneira) [reply Cabal/src/Distribution/Simple/Setup.hs:567]: > Lastly my understanding is that hie-bios currently always does at least one (implicit) build call anyway though the repl command so we're just maintaining the status quo if it has to do one with a special flag to get the buildinfo.

yeah but it would be nice if we get to make no one if possible, though

Comment (fendor) [reply Cabal/src/Distribution/Simple/Setup.hs]: changed

File: cabal-install/src/Distribution/Client/Setup.hs
Comment (fendor) [reply cabal-install/src/Distribution/Client/Setup.hs:487]: Sure but where?

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (fendor) [reply cabal-install/src/Distribution/Client/ProjectPlanOutput.hs]: Done

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:122]: Afaict, the allowed compiler programs for buildings are here: https://github.com/haskell/cabal/blob/496d6fcc26779e754523a6cc7576aea49ef8056e/Cabal/src/Distribution/Simple/Build.hs#L622
Looks like cabal would abort if it isn't one of these four compilers.

Comment (DanielG) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:122]: Right, but that doesn't imply that the `lookupProgram` call will succeed. I would expect the Simple.GHC module to have a bunch of similar "fail if ghc isn't in the ProgramDB" codepaths and indeed it does stuff like:

```
      let ghcProg = fromMaybe (error "GHC.pkgRoot: no ghc program") $ lookupProgram ghcProgram (withPrograms lbi)
```

However mostly it uses `requireProgram` instead which seems like what we want here too. It emits a nice error if the program can't be found. You have to call it in an IO context though.

As far as "the compiler can be found in the ProgramDB" being an invariant goes, at least the Simple.GHC module does a very early requireProgram call in `configure` so we can surmise that we wouldn't be able to get to the build stage if that doesn't hold.

Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:122]: So, we can either lift `mkBuildInfo` to `IO` or pass a `ConfiguredProgram` to `mkBuildInfo` or go with the, let it error approach.

I can find at least three calls to `fromMaybe ... lookupProgram ghcProgram` so it is likely to be safe here too.

File: Cabal/src/Distribution/Simple/Utils/Json.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/Utils/Json.hs]: Changes are merged independently!

File: Cabal/src/Distribution/Simple/BuildPaths.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/BuildPaths.hs:72]: I choose consistency with `debug-info`, everything will be renamed to `build-info`

File: doc/cabal-project.rst
Comment (DanielG) [inline doc/cabal-project.rst:1439]: s/human readable/machine readable/ ?

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1050]: Having the schema sure is useful but I think it would still be good to have a (perhaps abbreviated) example json here.

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1056]:
```diff
- Already generated build information will be removed.
+ Already generated `build-info.json` files will be removed since they would be stale otherwise. 
```

File: doc/cabal-project.rst
Comment (DanielG) [inline doc/cabal-project.rst:1447]: Exactly where the path can be found in plan.json seems undocumented still?

File: cabal-testsuite/PackageTests/Configure/include/HsZlibConfig.h.in
Comment (DanielG) [inline cabal-testsuite/PackageTests/Configure/include/HsZlibConfig.h.in:6]: Why is this (unrelated?) file being changed by this PR?

File: doc/cabal-project.rst
Comment (DanielG) [inline doc/cabal-project.rst:1437]: s/buildinfo/build-info/

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1037]: s/buildinfo/build-info/

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1052]: s/buildinfo/build-info/

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (DanielG) [inline cabal-install/src/Distribution/Client/ProjectPlanOutput.hs:189]: Is this version bound still correct? Shouldn't this be < 3.6?

**Review by DanielG — APPROVED** [review]
Looking very good. Some doc typos/inconsistencies are still left to fix but the implementation work is pretty much done.

I say this can be merged as soon as the docs are up to snuff and CI is green[1].

[1]: It's currently broken because a Cabal version bump is needed for the filterConfigureFlags logic to work, see #7571

File: doc/cabal-project.rst
Comment (fendor) [reply doc/cabal-project.rst:1447]: `plan.json` is undocumented in general, afaict. I did not really feel like documenting `plan.json`, too :) (But if that is what it takes to get this merged, so be it)

File: cabal-testsuite/PackageTests/Configure/include/HsZlibConfig.h.in
Comment (fendor) [reply cabal-testsuite/PackageTests/Configure/include/HsZlibConfig.h.in:6]: I have honestly no idea how this happened...
Will undo

File: doc/setup-commands.rst
Comment (fendor) [reply doc/setup-commands.rst:1084]: Added an example!

File: doc/cabal-project.rst
Comment (DanielG) [reply doc/cabal-project.rst:1447]: I'd be happy with just an approximate description of where the build-info field is. Something like:

> the exact location is available in the `"build-info"` field of each build component in the `"install-plan"` map in ``plan.json``.

(didn't check if that's correct though)

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (DanielG) [inline cabal-install/src/Distribution/Client/ProjectPlanOutput.hs:192]: I feel like it'd be better to have this be null instead of just missing unless you chose to do it this way for a particular reason?

File: doc/cabal-project.rst
Comment (DanielG) [reply doc/cabal-project.rst:1447]: Also this should explicitly document that the field can be missing/null when the Cabal version too old.

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (fendor) [reply cabal-install/src/Distribution/Client/ProjectPlanOutput.hs:192]: Of course, mental typo!

File: Cabal/src/Distribution/Simple/Build.hs
Comment (fendor) [inline Cabal/src/Distribution/Simple/Build.hs:120]: @DanielG, I moved this before building, afaict this allows us to extract the information, even if there is a compilation error. You think this is fine, too?

File: Cabal/src/Distribution/Simple/Build.hs
Comment (emilypi) [inline Cabal/src/Distribution/Simple/Build.hs:140]: These branches are mutually exclusive right? They shouldn't really exist in the same code block under a `when` imo - we should probably pull this out to an auxiliary function. It's important work!

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (emilypi) [inline Cabal/src/Distribution/Simple/ShowBuildInfo.hs:196]: The way we propagate errors in tuples here leaves alot to be desired from a code perspective imo. This makes me wonder whether we should introduce the idea of a warning-accumulating build state monad that can track these a little better. Not a blocker; just musing. I'd rather have this in than not!

**Review by emilypi — COMMENTED** [review]
Very nice @fendor! I'm impressed, and have nothing but nits, so consider this an approval if it's ready to go in

File: Cabal/src/Distribution/Simple/Build.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/Build.hs:140]: Good suggestion, will extract it!

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:196]: Yeah, the logic here is rather bad. But I would like to leave it for a future refactoring

**Comment by jneira** [comment]
do you think it is ready to merge @DanielG, @fgaz, @hasufell ?

File: doc/json-schemas/build-info.schema.json
Comment (hasufell) [inline doc/json-schemas/build-info.schema.json:5]: `cabal-install` version?

**Comment by jneira** [comment]
To add some more context:
* the json schema is captured in a tiny library, which can be used for producers and consumers (although the lib is not used here)
  * https://github.com/mpickering/hie-bios/issues/269#issuecomment-884344516
  * https://github.com/fendor/cabal-build-info/
* a good example of a consumer would be hie-bios itself (and ghcide/hls trough it):
  * https://github.com/mpickering/hie-bios/pull/301
  * https://github.com/haskell/haskell-language-server/pull/2009 (using the previous one)

File: doc/json-schemas/build-info.schema.json
Comment (fendor) [reply doc/json-schemas/build-info.schema.json:5]: no, this is intended to be the `lib:Cabal` version.

File: Cabal/src/Distribution/Simple/Build.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/Build.hs:122]: Done!

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs]: I decided to resolve this by moving a 'requireProgram' call to an outer function that has access to IO.

File: doc/json-schemas/build-info.schema.json
Comment (DanielG) [reply doc/json-schemas/build-info.schema.json:5]: plan.json has `"cabal-lib-version"` for that so we should probably be consistent with that here.

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1055]: this should also be explicit about this being the lib:Cabal version

File: doc/cabal-project.rst
Comment (DanielG) [reply doc/cabal-project.rst:1447]: This still doesn't seem to mention what the field name in plan.json is or where that field can be found like I suggested above.

File: cabal-install/src/Distribution/Client/ProjectPlanOutput.hs
Comment (DanielG) [inline cabal-install/src/Distribution/Client/ProjectPlanOutput.hs:184]: These docs need to be updated to reflect the fact that we null the field now instead of leaving it out.

File: doc/json-schemas/build-info.schema.json
Comment (fendor) [reply doc/json-schemas/build-info.schema.json:5]: Updated

File: Cabal/src/Distribution/Simple/Setup.hs
Comment (DanielG) [inline Cabal/src/Distribution/Simple/Setup.hs:280]: Haha! The build doesn't need to be successful anymore, does it :)

File: doc/setup-commands.rst
Comment (DanielG) [inline doc/setup-commands.rst:1082]: As discussed on IRC this should link to the code and the code should link here to prevent changes breaking the docs.

File: doc/json-schemas/build-info.schema.json
Comment (DanielG) [inline doc/json-schemas/build-info.schema.json:1]: The schema should also link to the code and the other way round.

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (DanielG) [inline Cabal/src/Distribution/Simple/ShowBuildInfo.hs:107]: The doc comment here should link to the docs and json schema to remind people to update those as well if changes are made here.

Something like:
> If you change this dont' forget to update the docs <here> and json schema <there>.

**Review by DanielG — COMMENTED** [review]
We should add some comments to make sure code, json schema and docs stay in sync in the face of other people changing this code in the future.

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (jneira) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:107]: or even the correlated lib representing the JSON parser

File: doc/setup-commands.rst
Comment (fendor) [reply doc/setup-commands.rst:1082]: Thanks, should be taken care of now!

File: Cabal/src/Distribution/Simple/ShowBuildInfo.hs
Comment (fendor) [reply Cabal/src/Distribution/Simple/ShowBuildInfo.hs:110]: Linked the other locations, but not the inofficial, unreleased encoder/decoder :P

File: doc/json-schemas/build-info.schema.json
Comment (fendor) [reply doc/json-schemas/build-info.schema.json:1]: Should be done

**Comment by jneira** [comment]
merge this

**Comment by jneira** [comment]
are you scared of pressing the button? :stuck_out_tongue_closed_eyes:

**Comment by fendor** [comment]
Well, someone else should press the merge button :)

**Comment by DanielG** [comment]
Alright then, been eyeing that button all day :D

**Comment by Mikolaj** [comment]
Uhoh. We *do* need `make users-guide` in CI.

**Comment by DanielG** [comment]
Indeed fendor seems to have forgotten to add the jsonschema extension to conf.py, but the users-guide target was already broken before the merge AFAICT :]

According to my bisect 0f041b603 is to blame originally.

**Comment by Mikolaj** [comment]
See #7637.

**Pull request merged**
